File: ABC177/D_Friends/submission1.java
// TLE

import java.util.Scanner;

class Main {
  public static void main(String[] args) {

    var sc = new Scanner(System.in);

    var n = sc.nextInt();
    var m = sc.nextInt();

    var ab = new int[2 * m];
    for (int i = 0; i < m; i++) {
      var a = sc.nextInt() - 1;
      var b = sc.nextInt() - 1;
      if (a > b) {
        var t = b;
        b = a;
        a = t;
      }
      for (int j = i; j >= 0; j--) {
        if (j == 0 || ab[2 * j - 2] <= a) {
          ab[2 * j    ] = a;
          ab[2 * j + 1] = b;
          break;
        }
        ab[2 * j    ] = ab[2 * j - 2];
        ab[2 * j + 1] = ab[2 * j - 1];
      }
    }

    var cluster = new int[n];
    var ic = 1;
    while (true) {
      int i = 0;
      for (; i < n; i++) {
        if (cluster[i] == 0) {
          cluster[i] = ic;
          break;
        }
      }
      if (i >= n) {
        break;
      }
      int iab = 0;
      int j = i;
      int nextJ = n;
      while (iab < m) {
        if (ab[2 * iab] < j) {
          iab++;
          continue;
        }
        if (ab[2 * iab] > j) {
          j = nextJ;
          nextJ = n;
          continue;
        }
        if (cluster[j] == ic) {
          int jj = ab[2 * iab + 1];
          cluster[jj] = ic;
          if (jj < nextJ) {
            nextJ = jj;
          }
        }
        iab++;
      }
      ic++;
    }

    var count = new int[ic];
    for (int i = 0; i < n; i++) {
      count[cluster[i]]++;
    }

    var max = 1;
    for (int i = 1; i < ic; i++) {
      var c = count[i];
      if (c > max) {
        max = c;
      }
    }

    System.out.println(max);

  }
}

File: ABC151/D_Maze_Master/submission1.scala
// AC

object Main extends App {
  val sc = new java.util.Scanner(System.in);
  val h, w = sc.nextInt();
  sc.nextLine(); // 行末スキップ
  val matrix = Array.fill(h)(sc.nextLine().map(ch => if(ch=='#') -1 else 0).toArray);

  def resetMatrix(): Unit = {
    (0 until h).foreach { y =>
      (0 until w).foreach { x =>
        if (matrix(y)(x) > 0) {
          matrix(y)(x) = 0;
        }
      }
    }
  }

  def maxDistance(x: Int, y: Int): Int = {
    @scala.annotation.tailrec
    def sub1(que: List[(Int, Int)], distance: Int): Int = {
      if (que.isEmpty) {
        distance - 2;
      } else {
        val que2 = sub2(que, Nil, distance);
        sub1(que2, distance + 1);
      }
    }
    @scala.annotation.tailrec
    def sub2(que1: List[(Int, Int)], que2: List[(Int, Int)], distance: Int): List[(Int, Int)] = {
      que1 match {
        case Nil => que2;
        case head :: q =>
          val x = head._1;
          val y = head._2;
          val que3 = List((x-1, y), (x, y-1), (x+1, y), (x, y+1)).filter { case (x, y) =>
            if (x >= 0 && x < w && y >= 0 && y < h) {
              if (matrix(y)(x) == 0) {
                matrix(y)(x) = distance + 1;
                true;
              } else {
                false;
              }
            } else {
              false;
            }
          } ::: que2;
          sub2(q, que3, distance);
      }
    }
    matrix(y)(x) = 1;
    sub1(List((x, y)), 1);
  }

  val answer = (0 until h).map { y =>
    (0 until w).map { x =>
      if (matrix(y)(x) >= 0) {
        resetMatrix();
        maxDistance(x, y);
      } else {
        -1;
      }
    }.max;
  }.max;

  println(answer);
}
